Mathlib/Topology/UnitInterval.lean

```diff
@@ -291,6 +291,126 @@ lemma abs_sub_addNSMul_le (hδ : 0 ≤ δ) {t : Icc a b} (n : ℕ)
           rw [add_sub_add_comm, sub_self, zero_add, succ_nsmul', add_sub_cancel_right]
           exact (abs_eq_self.mpr hδ).le
 
+/--
+Form a convex linear combination of two points in a closed interval.
+
+This should be removed once a general theory of convex spaces is available in Mathlib.
+-/
+def convexCombo {a b : ℝ} (x y : Icc a b) (t : unitInterval) : Icc a b :=
+  ⟨(1 - t) * x + t * y, by
+    constructor
+    · nlinarith [x.2.1, y.2.1, t.2.1, t.2.2]
+    · nlinarith [x.2.2, y.2.2, t.2.1, t.2.2]⟩
+
+@[simp, grind =]
+theorem coe_convexCombo {a b : ℝ} (x y : Icc a b) (t : unitInterval) :
+  (convexCombo x y t : ℝ) = (1 - t) * x + t * y := rfl
+
+@[simp, grind =]
+theorem convexCombo_zero {a b : ℝ} (x y : Icc a b) : convexCombo x y 0 = x := by
+  simp [convexCombo]
+
+@[simp, grind =]
+theorem convexCombo_one {a b : ℝ} (x y : Icc a b) : convexCombo x y 1 = y := by
+  simp [convexCombo]
+
+@[simp, grind =]
+theorem convexCombo_symm {a b : ℝ} (x y : Icc a b) (t : unitInterval) :
+    convexCombo x y (unitInterval.symm t) = convexCombo y x t := by
+  simp [convexCombo]
+  abel
+
+@[grind .]
+theorem le_convexCombo {a b : ℝ} {x y : Icc a b} (h : x ≤ y) (t : unitInterval) :
+    x ≤ convexCombo x y t := by
+  change (x : ℝ) ≤ _
+  change (x : ℝ) ≤ _ at h
+  simp [convexCombo]
+  nlinarith [t.2.1, t.2.2]
+
+@[grind .]
+theorem convexCombo_le {a b : ℝ} {x y : Icc a b} (h : x ≤ y) (t : unitInterval) :
+    convexCombo x y t ≤ y := by
+  change _ ≤ (y : ℝ)
+  change (x : ℝ) ≤ _ at h
+  simp [convexCombo]
+  nlinarith [t.2.1, t.2.2]
+
+/--
+Helper definition for `convexCombo_assoc`, giving one of the coefficients appearing
+when we reassociate a convex combination.
+-/
+abbrev convexCombo_assoc_coeff₁ (s t : unitInterval) : unitInterval :=
+  ⟨s * (1 - t) / (1 - s * t),
+    by
+      apply div_nonneg
+      · nlinarith [s.2.1, t.2.2]
+      · nlinarith [s.2.2, t.2.2, t.2.1],
+    by
+      apply div_le_one_of_le₀
+      · nlinarith [s.2.2]
+      · nlinarith [s.2.2, t.2.2, t.2.1]⟩
+
+/--
+Helper definition for `convexCombo_assoc`, giving one of the coefficients appearing
+when we reassociate a convex combination.
+-/
+abbrev convexCombo_assoc_coeff₂ (s t : unitInterval) : unitInterval := s * t
+
+theorem convexCombo_assoc {a b : ℝ} (x y z : Icc a b) (s t : unitInterval) :
```

Comment:
Do you think it's useful to also have a variant for rewriting in the other direction? I.e., with `assoc_coeff`s on the left.

Reply:
It's probably not that useful having either, as the useful operation is n-ary convex combination... but it's cheap to add.